Bucketeer/Sources/Internal/Database/SQLite/SQLite.swift

```diff
@@ -5,31 +5,37 @@ final class SQLite {
     let path: String
     let pointer: OpaquePointer
     let logger: Logger?
+    // access SQLite on one serial queue to prevent `database is locked` or database corrupt
+    private static let dbQueue = DispatchQueue(label: "io.bucketeer.SQLite")
 
     init(path: String, logger: Logger?) throws {
         self.path = path
         self.logger = logger
-
-        sqlite3_shutdown()
-        sqlite3_initialize()
-
-        var _pointer: OpaquePointer?
-        let flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX
-        let result = sqlite3_open_v2(path, &_pointer, flags, nil)
-        if result != SQLITE_OK {
-            guard let pointer = _pointer else {
+        pointer = try Self.dbQueue.sync {
+            var _pointer: OpaquePointer?
+            let flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX
+            let result = sqlite3_open_v2(path, &_pointer, flags, nil)
+            if result != SQLITE_OK {
+                guard let dbConnection = _pointer else {
+                    throw Error.unknown
+                }
+                throw Error.failedToOpen(.init(errorMessage: dbConnection.readErrorMessage(), result: result))
+            }
+            guard let dbConnection = _pointer else {
                 throw Error.unknown
             }
-            throw Error.failedToOpen(.init(pointer: pointer, result: result))
-        }
-        guard let pointer = _pointer else {
-            throw Error.unknown
+            return dbConnection
         }
-        self.pointer = pointer
+        // allow concurrent reads and writes https://www.sqlite.org/walformat.html
+        // its nice to have
+        try exec(query: "PRAGMA journal_mode = WAL")
```

Comment:
WAL, although comment on this line the SDK is still working.
I am thinking: should we keep it?
Because we only have one database connection for now.

Reply:
It's not necessary to support concurrent operations. Also, we are using the serial queue, so it makes sense to leave it as it was before.

Reply:
I will remove it

Reply:
removed https://github.com/bucketeer-io/ios-client-sdk/pull/40/commits/200675addeca080bc2e9f34bfb315662aceaefad